File: packages/core/test/otlp.test.ts
import { describe, expect, it } from "vitest";
import { extractClaudeMetrics, parseGithubHints } from "../src/otlp.js";
import { otlpMetricsRequestSchema } from "../src/schemas/otlp.js";

const kv = (key: string, stringValue: string) => ({ key, value: { stringValue } });

function tokenPoint(type: string, value: number, sessionId = "sess-1") {
  return {
    attributes: [
      kv("session.id", sessionId),
      kv("type", type),
      kv("model", "claude-sonnet-5"),
    ],
    asDouble: value,
    timeUnixNano: "1752994800000000000",
  };
}

const samplePayload = {
  resourceMetrics: [
    {
      resource: {
        attributes: [
          kv("service.name", "claude-code"),
          kv("github.repo", "acme/api"),
          kv("github.run_id", "16123456789"),
          kv("github.pr_number", "42"),
        ],
      },
      scopeMetrics: [
        {
          scope: { name: "com.anthropic.claude_code" },
          metrics: [
            {
              name: "claude_code.token.usage",
              sum: {
                dataPoints: [
                  tokenPoint("input", 1200),
                  tokenPoint("output", 340),
                  tokenPoint("cacheRead", 9000),
                ],
                aggregationTemporality: 2,
                isMonotonic: true,
              },
            },
            {
              name: "claude_code.cost.usage",
              sum: {
                dataPoints: [
                  {
                    attributes: [kv("session.id", "sess-1"), kv("model", "claude-sonnet-5")],
                    asDouble: 0.42,
                  },
                ],
                aggregationTemporality: 2,
                isMonotonic: true,
              },
            },
            {
              name: "some_other.metric",
              sum: { dataPoints: [{ asDouble: 1 }] },
            },
          ],
        },
      ],
    },
  ],
};

describe("extractClaudeMetrics", () => {
  it("extracts claude_code points with session ids, drops the rest", () => {
    const parsed = otlpMetricsRequestSchema.parse(samplePayload);
    const batches = extractClaudeMetrics(parsed);
    expect(batches).toHaveLength(1);
    const batch = batches[0]!;
    expect(batch.resourceAttrs["github.repo"]).toBe("acme/api");
    expect(batch.points).toHaveLength(4);
    const input = batch.points.find((p) => p.attrType === "input")!;
    expect(input.value).toBe(1200);
    expect(input.temporality).toBe("cumulative");
    expect(input.sessionId).toBe("sess-1");
  });

  it("handles asInt values encoded as strings (proto3 JSON int64)", () => {
    const payload = otlpMetricsRequestSchema.parse({
      resourceMetrics: [
        {
          scopeMetrics: [
            {
              metrics: [
                {
                  name: "claude_code.token.usage",
                  sum: {
                    dataPoints: [
                      {
                        attributes: [kv("session.id", "s"), kv("type", "input")],
                        asInt: "5000",
                      },
                    ],
                    aggregationTemporality: 1,
                  },
                },
              ],
            },
          ],
        },
      ],
    });
    const [batch] = extractClaudeMetrics(payload);
    expect(batch!.points[0]!.value).toBe(5000);
    expect(batch!.points[0]!.temporality).toBe("delta");
  });

  it("drops data points without a session.id", () => {
    const payload = otlpMetricsRequestSchema.parse({
      resourceMetrics: [
        {
          scopeMetrics: [
            {
              metrics: [
                {
                  name: "claude_code.token.usage",
                  sum: { dataPoints: [{ asDouble: 10, attributes: [kv("type", "input")] }] },
                },
              ],
            },
          ],
        },
      ],
    });
    expect(extractClaudeMetrics(payload)).toHaveLength(0);
  });
});

describe("parseGithubHints", () => {
  it("parses run id and pr number as numbers", () => {
    const hints = parseGithubHints({
      "github.repo": "acme/api",
      "github.run_id": "16123456789",
      "github.pr_number": "42",
    });
    expect(hints).toEqual({ repoFullName: "acme/api", ghRunId: 16123456789, prNumber: 42 });
  });

  it("ignores malformed values", () => {
    expect(parseGithubHints({ "github.run_id": "not-a-number" })).toEqual({});
  });
});

File: internal/otlp/otlp_test.go
package otlp

import (
	"encoding/json"
	"testing"
)

func parse(t *testing.T, raw string) *Request {
	t.Helper()
	var req Request
	if err := json.Unmarshal([]byte(raw), &req); err != nil {
		t.Fatalf("parse: %v", err)
	}
	return &req
}

const samplePayload = `{
  "resourceMetrics": [{
    "resource": {"attributes": [
      {"key": "service.name", "value": {"stringValue": "claude-code"}},
      {"key": "github.repo", "value": {"stringValue": "acme/api"}},
      {"key": "github.run_id", "value": {"stringValue": "16123456789"}},
      {"key": "github.pr_number", "value": {"stringValue": "42"}}
    ]},
    "scopeMetrics": [{
      "metrics": [
        {"name": "claude_code.token.usage", "sum": {
          "aggregationTemporality": 2, "isMonotonic": true,
          "dataPoints": [
            {"asDouble": 1200, "attributes": [{"key":"session.id","value":{"stringValue":"sess-1"}},{"key":"type","value":{"stringValue":"input"}},{"key":"model","value":{"stringValue":"claude-sonnet-5"}}]},
            {"asDouble": 340, "attributes": [{"key":"session.id","value":{"stringValue":"sess-1"}},{"key":"type","value":{"stringValue":"output"}},{"key":"model","value":{"stringValue":"claude-sonnet-5"}}]},
            {"asDouble": 9000, "attributes": [{"key":"session.id","value":{"stringValue":"sess-1"}},{"key":"type","value":{"stringValue":"cacheRead"}},{"key":"model","value":{"stringValue":"claude-sonnet-5"}}]}
          ]}},
        {"name": "claude_code.cost.usage", "sum": {
          "aggregationTemporality": 2,
          "dataPoints": [{"asDouble": 0.42, "attributes": [{"key":"session.id","value":{"stringValue":"sess-1"}},{"key":"model","value":{"stringValue":"claude-sonnet-5"}}]}]}},
        {"name": "some_other.metric", "sum": {"dataPoints": [{"asDouble": 1}]}}
      ]
    }]
  }]
}`

func TestExtract(t *testing.T) {
	batches := Extract(parse(t, samplePayload))
	if len(batches) != 1 {
		t.Fatalf("want 1 batch, got %d", len(batches))
	}
	b := batches[0]
	if b.ResourceAttrs["github.repo"] != "acme/api" {
		t.Errorf("resource attr lost: %v", b.ResourceAttrs)
	}
	if len(b.Points) != 4 {
		t.Fatalf("want 4 points, got %d", len(b.Points))
	}
	var input *MetricPoint
	for i := range b.Points {
		if b.Points[i].AttrType == "input" {
			input = &b.Points[i]
		}
	}
	if input == nil || input.Value != 1200 || input.Temporality != Cumulative || input.SessionID != "sess-1" {
		t.Errorf("input point wrong: %+v", input)
	}
}

func TestExtractStringAsInt(t *testing.T) {
	raw := `{"resourceMetrics":[{"scopeMetrics":[{"metrics":[
		{"name":"claude_code.token.usage","sum":{"aggregationTemporality":1,
		"dataPoints":[{"asInt":"5000","attributes":[{"key":"session.id","value":{"stringValue":"s"}},{"key":"type","value":{"stringValue":"input"}}]}]}}]}]}]}`
	batches := Extract(parse(t, raw))
	if len(batches) != 1 || len(batches[0].Points) != 1 {
		t.Fatal("expected one point")
	}
	p := batches[0].Points[0]
	if p.Value != 5000 || p.Temporality != Delta {
		t.Errorf("got %+v", p)
	}
}

func TestExtractDropsSessionlessPoints(t *testing.T) {
	raw := `{"resourceMetrics":[{"scopeMetrics":[{"metrics":[
		{"name":"claude_code.token.usage","sum":{"dataPoints":[{"asDouble":10,"attributes":[{"key":"type","value":{"stringValue":"input"}}]}]}}]}]}]}`
	if got := Extract(parse(t, raw)); len(got) != 0 {
		t.Errorf("sessionless point survived: %+v", got)
	}
}

func TestParseGithubHints(t *testing.T) {
	h := ParseGithubHints(map[string]string{
		"github.repo":      "acme/api",
		"github.run_id":    "16123456789",
		"github.pr_number": "42",
	})
	if h.RepoFullName != "acme/api" || h.GhRunID != 16123456789 || h.PRNumber != 42 {
		t.Errorf("got %+v", h)
	}
	if bad := ParseGithubHints(map[string]string{"github.run_id": "not-a-number"}); bad.GhRunID != 0 {
		t.Errorf("malformed run id accepted: %+v", bad)
	}
}
